File: python/vedadb_orm/graph.py
"""
VedaDB ORM — GraphProxy.

Provides a model-scoped interface to VedaDB's graph-engine features:
node/edge management, traversal, shortest-path, and PageRank.
"""

from __future__ import annotations

import json
from typing import Any, Dict, List, Optional, TYPE_CHECKING

from .exceptions import QueryError
from .utils import format_where_value

if TYPE_CHECKING:
    pass


class GraphProxy:
    """Model-scoped graph operations."""

    def __init__(self, model_cls: type) -> None:
        self._model_cls = model_cls

    def _orm(self) -> Any:
        orm = self._model_cls._get_orm()
        if orm is None:
            raise QueryError("Model is not bound to a VedaORM instance")
        return orm

    @property
    def _table(self) -> str:
        return self._model_cls._table_name

    # -- node / edge management ---------------------------------------------

    def add_node(self, id: Any, label: str, **props: Any) -> None:
        """Insert a graph node."""
        props_json = json.dumps(props) if props else "{}"
        sql = (
            f"GRAPH ADD NODE {self._table} "
            f"ID {format_where_value(id)} "
            f"LABEL '{label}' "
            f"PROPERTIES '{props_json}';"
        )
        self._orm().query(sql)

    def add_edge(
        self,
        from_id: Any,
        to_id: Any,
        edge_type: str,
        **props: Any,
    ) -> None:
        """Insert a directed edge between two nodes."""
        props_json = json.dumps(props) if props else "{}"
        sql = (
            f"GRAPH ADD EDGE {self._table} "
            f"FROM {format_where_value(from_id)} "
            f"TO {format_where_value(to_id)} "
            f"TYPE '{edge_type}' "
            f"PROPERTIES '{props_json}';"
        )
        self._orm().query(sql)

    # -- traversal ----------------------------------------------------------

    def traverse(
        self,
        start: Any,
        depth: int = 1,
        edge_type: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Breadth-first traversal from *start* up to *depth* hops.

        Returns a list of node dicts.
        """
        sql = (
            f"GRAPH TRAVERSE {self._table} "
            f"FROM {format_where_value(start)} "
            f"DEPTH {depth}"
        )
        if edge_type:
            sql += f" EDGE_TYPE '{edge_type}'"
        sql += ";"

        result = self._orm().query(sql)
        return [dict(zip(result.columns, row)) for row in result.rows]

    def shortest_path(self, start: Any, end: Any) -> List[Dict[str, Any]]:
        """Find the shortest path between two nodes.

        Returns a list of node dicts representing the path.
        """
        sql = (
            f"GRAPH SHORTEST_PATH {self._table} "
            f"FROM {format_where_value(start)} "
            f"TO {format_where_value(end)};"
        )
        result = self._orm().query(sql)
        return [dict(zip(result.columns, row)) for row in result.rows]

    def pagerank(self) -> List[Dict[str, Any]]:
        """Run PageRank on the graph and return ranked nodes.

        Returns a list of dicts with at least ``id`` and ``rank`` keys.
        """
        sql = f"GRAPH PAGERANK {self._table};"
        result = self._orm().query(sql)
        return [dict(zip(result.columns, row)) for row in result.rows]

File: node/vedadb-orm/src/graph.ts
/**
 * VedaDB ORM GraphProxy
 *
 * Provides graph traversal capabilities via VedaDB GRAPH commands.
 */

import { DriverResult } from './types';
import { SchemaDefinition } from './schema';
import { escapeIdentifier, escapeValue } from './utils';

export interface GraphTraversalOptions {
  direction?: 'OUT' | 'IN' | 'BOTH';
  edgeType?: string;
  maxDepth?: number;
  limit?: number;
}

export class GraphProxy<T = any> {
  private schema: SchemaDefinition<T>;
  private executor: (sql: string) => Promise<DriverResult>;
  private hydrator: (result: DriverResult) => T[];

  constructor(
    schema: SchemaDefinition<T>,
    executor: (sql: string) => Promise<DriverResult>,
    hydrator: (result: DriverResult) => T[]
  ) {
    this.schema = schema;
    this.executor = executor;
    this.hydrator = hydrator;
  }

  /**
   * Add an edge between two nodes.
   */
  async addEdge(fromId: any, toId: any, edgeType: string, properties?: Record<string, any>): Promise<void> {
    let sql = `GRAPH ADD EDGE '${edgeType}' FROM ${escapeValue(fromId)} TO ${escapeValue(toId)}`;
    if (properties && Object.keys(properties).length > 0) {
      const props = Object.entries(properties)
        .map(([k, v]) => `${escapeIdentifier(k)} = ${escapeValue(v)}`)
        .join(', ');
      sql += ` SET ${props}`;
    }
    sql += ';';
    await this.executor(sql);
  }

  /**
   * Remove an edge between two nodes.
   */
  async removeEdge(fromId: any, toId: any, edgeType: string): Promise<void> {
    await this.executor(
      `GRAPH REMOVE EDGE '${edgeType}' FROM ${escapeValue(fromId)} TO ${escapeValue(toId)};`
    );
  }

  /**
   * Traverse the graph from a starting node.
   */
  async traverse(startId: any, options: GraphTraversalOptions = {}): Promise<T[]> {
    const table = escapeIdentifier(this.schema.tableName);
    const direction = options.direction || 'OUT';
    const maxDepth = options.maxDepth || 3;
    const limit = options.limit || 100;

    let sql = `GRAPH TRAVERSE ${table} FROM ${escapeValue(startId)} DIRECTION ${direction} DEPTH ${maxDepth}`;
    if (options.edgeType) sql += ` EDGE '${options.edgeType}'`;
    sql += ` LIMIT ${limit};`;

    const result = await this.executor(sql);
    return this.hydrator(result);
  }

  /**
   * Find shortest path between two nodes.
   */
  async shortestPath(fromId: any, toId: any, edgeType?: string): Promise<T[]> {
    const table = escapeIdentifier(this.schema.tableName);
    let sql = `GRAPH SHORTEST_PATH ${table} FROM ${escapeValue(fromId)} TO ${escapeValue(toId)}`;
    if (edgeType) sql += ` EDGE '${edgeType}'`;
    sql += ';';

    const result = await this.executor(sql);
    return this.hydrator(result);
  }

  /**
   * Get neighbors of a node.
   */
  async neighbors(id: any, options: GraphTraversalOptions = {}): Promise<T[]> {
    return this.traverse(id, { ...options, maxDepth: 1 });
  }
}
